sklearn/cluster/tests/test_hierarchical.py

```diff
@@ -559,3 +594,51 @@ def increment(self, *args, **kwargs):
     linkage_tree(X, connectivity=connectivity, affinity=fa.increment)
 
     assert_equal(fa.counter, 3)
+
+
+def test_agglomerative_clustering_with_distance_threshold():
+    # Check that we obtain the correct number of clusters with
+    # agglomerative clustering with distance_threshold.
+
+    rng = np.random.RandomState(0)
+    mask = np.ones([10, 10], dtype=np.bool)
+    n_samples = 100
+    X = rng.randn(n_samples, 50)
+    connectivity = grid_to_graph(*mask.shape)
+    # test when distance threshold is set to 10
+    distance_threshold = 10
+    for linkage in ("ward", "complete", "average"):
+        for conn in [None, connectivity]:
+            clustering = AgglomerativeClustering(
+                distance_threshold=distance_threshold,
+                connectivity=conn, linkage=linkage)
+            clustering.fit(X)
+            clusters_produced = clustering.labels_
+            num_clusters_produced = len(np.unique(clustering.labels_))
+            # test if the clusters produced match the point in the linkage tree
+            # where the distance exceeds the threshold
+            tree_builder = _TREE_BUILDERS[linkage]
+            children, n_components, n_leaves, parent, distances = \
+                tree_builder(X, connectivity=conn, n_clusters=None,
+                             return_distance=True)
+            num_clusters_at_threshold = np.count_nonzero(
```

Comment:
Perhaps we should test something more explicit, just to be sure that your logic here is correct, like check that in single linkage, the minimum within-cluster pairwise distance for each sample is under the threshold and the minimum out-of-cluster pairwise distance is greater.

Reply:
Do you mean along the lines of [this](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/cluster/tests/test_hierarchical.py#L326) test? 
I could use the same dataset to do a more explicit test.

Reply:
I don't see how it relates to that test. I mean that for some `X` and some predicted `labels`:
```
D = pairwise_distances(X, metric=metric)
for i in range(len(X)):
    in_cluster_mask = labels == labels[i]
    max_in_cluster_distance = D[i, in_cluster_mask].max()
    min_out_cluster_distance = D[i, ~in_cluster_mask].min()
    # XXX: there should be equality on one of these conditions
    assert max_in_cluster_distance < threshold
    assert min_in_cluster_distance > threshold
```

Reply:
Apologies for the delayed response, but as far as I understand the pairwise distance will give the distance between each point in `X` <b>not</b> the distance between the clusters as they join up. The distance between each cluster is in the `distances` matrix calculated using the `scipy.cluster.hierarchy.linkage` method.

So is there still a need to have an explicit test?

Reply:
True. But surely a similar invariance could be constructed about the average distances with average linkage...? I've not thought about it too rigorously.

Reply:
Although what I said is true only when `connectivity` is `None`. If a `connectivity` matrix is passed in then calculating the clusters would mean deciphering what is happening in the `ward_tree` and `linkage_tree` methods, and I'm not sure it's worth the effort...

>True. But surely a similar invariance could be constructed about the average distances with average linkage...? I've not thought about it too rigorously.

I suppose, but is there a need to do this? 
I'm really not sure how to do a more explicit test, so I'd really appreciate help with this.